docs/rfcs/040-neon-supported-extensions.md

```diff
@@ -0,0 +1,140 @@
+
+
+2. When compute is starting, cplane collects all extensions that are available for this compute image version and adds their `neon_supported_extension.id`, `neon_supported_extension.control` and `neon_supported_extension.library_index`, down to compute spec. 
+
+    To find available extensions for compute with `$current_image_version` and `$current_major_version`:
+
+    ```
+    SELECT id, control, library_index
+    FROM neon_supported_extension
+    WHERE  $current_major_version = compute_major_version -- for multi-version compute images, this will be an array of versions
+    AND is_available_for_all
+    AND  $current_image_version >= compute_image_version
+    ORDER BY (compute_image_version, id) DESC
+    LIMIT 1;
+    ```
+
+    Q: do we really need control file in the compute spec? Can we just add them as files to compute image when image is built?
+
+    To override extension version (or enable private extension) for a specific project/user, we can add a column to the project/user table, i.e. `extension_ids array references neon_supported_extension(id)`. This column will replace currently existing `remote_extensions` part of `pg_settings`.
+
+3. Inside compute image, we will build path to custom extensions s3 based on provided spec and download them into the respective library_path locasion in compute image.
+    Almost the same way as we do now, but use `s3://neon-extensions/<neon_supported_extension.id>` as archive path.
+
+4. When `CREATE EXTENSION myextension;` is called, we will intercept the request to the file and load the extension from the custom path, based on compute version and extension name. This will allow support of multi-version compute images. This is also a good place to track extension usage statistics.
+
+    Alternatively we can just symlink `/usr/local/pgsql/lib/extension/<neon_supported_extension.id>/*` files to default path `/usr/local/pgsql/lib/` as soon as we know, which version of postgres is runnnig. This approach doesn't require postgres patch. Will it be enough to support multiple versions of extenension?
+
+5. There is a special case of backward incompatibe minor version upgrades of extension that use the same library name. I.e. `my_ext--1.0` and `my_ext--1.1` both use `my_ext.so` and version 1.1 drops some function that was present in 1.0. This is a bad practice and shouldn't happen often.
+
+    In this case, we will have to add special rule to `download_extension_file_hook()` to check installed version of extension and map requested `my_ext.so` to `my_ext-1.0.so` or `my_ext-1.1.so`. Build extension step must be aware of this change too - it will add suffix to extension library file and reflect this in `library_index[]`.
+
+    Note that this will also work if an extension is actually a pack of extensions with different versions (i.e. `postgis 3.3` that includes `postgis_sfcgal v 1.3.8`).
+
+    TODO: Find at least one example of such backward incompatible extension.
```

Comment:
https://github.com/pgpartman/pg_partman/tree/master

It doesn't provide old --sql files. And library can be incompatible as well, need to test it.